runtime/debug/strategy-explorer-adapter.js

```diff
@@ -22,9 +22,9 @@ export class StrategyExplorerAdapter {
     const idMap = new Map(); // Recipe -> ID
     let lastID = 0;
     const assignIdAndCopy = recipe => {
-      idMap.set(recipe, lastID);
-      const {result, score, derivation, description, hash, valid, active, irrelevant} = recipe;
-      return {result, score, derivation, description, hash, valid, active, irrelevant, id: lastID++};
+      idMap.set(JSON.stringify(recipe), lastID);
```

Comment:
It seems to me you've run into the issue of serializing the generations from the plan producer, where generations are a complex graph of objects that cannot be serialized easily - that's why you implemented the `formatSerializableGenerations` method.

This code here i.e. StrategyExplorerAdapter. adapt() is primarily dealing with this exact issue - assigning each recipe an ID and referring to rewriting the object references to references to IDs.

`formatSerializableGenerations` means that this code now has to do JSON.stringify() serialization to assign the IDs, which is more costly. And also, more importantly, we now have 2 places in code attempting to solve the same problem - both of them now doing it not as well as we've done previously.

I would propose that we should split the functionality exposed by `StrategyExplorerAdapter` into 2 standalone functions:
1) Transforming generations into a serializable data structure.
2) Sending the serialized data structure to DevTools.

Then we should use (1) in the plan producer, so that a serialized data structure can be put in firebase. We would read it in the plan consumer and, if devtools is connected, sent over to the DevtoolsConnection.

Sorry for being a bit strict here, but I wouldn't like us to introduce more technical debt while we're catching up on fixing it just now.

Reply:
Agreed, I would be happy to move the logic of serialization of generation to plan production.

This logic should be generic enough to serve multiple customers who would like to consume history and/or details of strategization. DevTools is just one of them.